core/jvm/src/main/scala/fs2/interop/flow/package.scala

```diff
@@ -0,0 +1,253 @@
+    *                   like requesting elements from a database.
+    *                   The publisher can use this `bufferSize` to query elements in batch.
+    *                   A high number will also lead to more elements in memory.
+    * @param onSubscribe The effectual function to run during the subscription process,
+    *                    receives a [[Subscriber]] that should be used to subscribe to a [[Publisher]].
+    *                    The `subscribe` operation must be called exactly once.
+    */
+  def fromSubscriber[F[_], A](
+      bufferSize: Int
+  )(
+      subscribe: Subscriber[A] => F[Unit]
+  )(implicit
+      F: Async[F]
+  ): Stream[F, A] =
+    Stream
+      .eval(StreamSubscriber[F, A](bufferSize))
+      .flatMap { subscriber =>
+        subscriber.stream(
+          F.uncancelable(_ => subscribe(subscriber))
+        )
+      }
+
+  /** Creates a [[Stream]] from an `onSubscribe` effectual function.
+    *
+    * This function is useful when you need to perform some effectual actions during the subscription process.
+    * The subscription process is uncancelable.
+    *
+    * @example {{{
+    * scala> import cats.effect.IO
+    * scala> import fs2.Stream
+    * scala> import fs2.interop.flow.fromOnSubscribe
+    * scala> import java.util.concurrent.Flow.Publisher
+    * scala>
+    * scala> def consumeThirdPartyPublisher[A](publisher: Publisher[A]): Stream[IO, A] =
+    *   fromOnSubscribe[IO, A](bufferSize = 16) { subscribe =>
+    *     IO.println("Subscribing!") >>
+    *     subscribe(publisher) >>
+    *     IO.println("Subscribed!")
+    *   }
+    * }}}
+    *
+    * @note The publisher only receives a subscriber when the stream is run.
+    *
+    * @see [[fromPublisher]] for a simpler version that only requires a [[Publisher]].
+    *
+    * @param bufferSize setup the number of elements asked each time from the [[Publisher]].
+    *                   A high number can be useful if the publisher is triggering from IO,
+    *                   like requesting elements from a database.
+    *                   The publisher can use this `bufferSize` to query elements in batch.
+    *                   A high number will also lead to more elements in memory.
+    * @param onSubscribe The effectual action to run during the subscription process,
+    *                    represented as a function that receives as an argument
+    *                    the actual `subscribe` operation; as another function receiving the [[Publisher]].
+    *                    The `subscribe` operation must be called exactly once.
+    */
+  def fromOnSubscribe[F[_], A](
+      bufferSize: Int
+  )(
+      onSubscribe: (Publisher[A] => F[Unit]) => F[Unit]
+  )(implicit
+      F: Async[F]
+  ): Stream[F, A] =
+    fromSubscriber[F, A](bufferSize) { subscriber =>
+      val subscribe: Publisher[A] => F[Unit] =
+        publisher => F.delay(publisher.subscribe(subscriber))
+
+      onSubscribe(subscribe)
+    }
```

Comment:
Should we expose another version that gives access to `Poll` to users?

Reply:
Tbh I'm confused about this API.
- `(Publisher[A] => F[Unit]) => F[Unit]` is a complex signature
- it doesn't represent any existing concept from the reactive streams spec AFAICT

Reply:
> (Publisher[A] => F[Unit]) => F[Unit] is a complex signature

IMHO, is very similar to `async` you pass a function that receives as an argument some kind of callback; thus another function.

> it doesn't represent any existing concept from the reactive streams spec AFAICT

Not directly, yes.
It is an effectual inversion of control of `Publisher.subscribe`.

The idea of this API is to be mostly as useful as `fromSubscriber` but with less surface for error, since the actual `Subscriber` is not exposed, only its capability to `subscribe` to a `Publisher`.
Having said that, one can indeed argue this offer little value when we already have the safe version and the fully flexible version.